### PR TITLE
feat: create-config

### PR DESCRIPTION
**Pull request by almog8k**
| Question                | Answer                                                                          |
| ---------------- | -------------------------------------------------------------------------- |
| Bug fix         | ✖                                                                        |
| New feature     | ✔                                                                       |
| Breaking change | ✖                                                                        |
| Deprecations    | ✖                                                                        |
| Documentation   | ✖                                                                        |
| Tests added     | ✖                                                                        |
| Chore            | ✖                                                                       |

Further  information:
Adding the Create Config feature (three step processes):
Step 1: Config Info (config metadata: configName, schemaId, version and etc... ) Form.
step 2: Config Data (config json value, using monacho for implementation).
step 3: Review and Approve : TBD

### CODE REVIEW

File: src/pages/createConfig/step2AddConfig/step2AddConfig.tsx [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+  const fetchSchema = useCallback(() => getSchema({ id: schemaId, shouldDereference: true }), [schemaId]);
+
+  const { data: schema, isSuccess } = useQuery({
+    queryKey: ['getSchema2'],
+    queryFn: fetchSchema,
+  });
+
+  const monaco = useMonaco();
+
+  const ajvInstance = useMemo(() => new ajv(), []);
```
Comment (CptSchnitz) [inline src/pages/createConfig/step2AddConfig/step2AddConfig.tsx:28]: you can initialise it outside of the component

**Review by CptSchnitz — CHANGES_REQUESTED** [review]

File: package.json [hunk 2]
```diff
@@ -23,10 +24,13 @@
     "@rjsf/utils": "^5.18.4",
     "@rjsf/validator-ajv8": "^5.18.4",
     "@tanstack/react-query": "^5.40.0",
+    "ajv": "^6.12.6",
```
[the remark this answers is not in the document]
Comment (almog8k) [reply package.json:27]: done

File: src/components/SchemaSelect/schemaSelect.types.ts [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+        options.push({
+          title: item.name,
+          id: item.id,
+          group: parentName,
+        });
+      }
+    }
+  });
+  return options;
+};
```
[the remark this answers is not in the document]
Comment (almog8k) [reply src/components/SchemaSelect/schemaSelect.types.ts:28]: done

File: src/components/SchemaSelect/schemaSelect.tsx [hunk 4]
```diff
@@ -0,0 +1,85 @@
+import { Autocomplete, Box, ListSubheader, TextField } from '@mui/material';
+import { useQuery } from '@tanstack/react-query';
+import { getSchemasTree, schemaTreeDir } from '../../api/client';
+import { useState } from 'react';
+import { GroupOption, flattenData } from './schemaSelect.types';
+import { SchemaSelectValue } from '../../pages/createConfig/step1GeneralInfo/step1GeneralInfo.schemas';
+
+type SchemaSelectProps = {
+  onChange: (value: SchemaSelectValue) => void;
```
[the remark this answers is not in the document]
Comment (almog8k) [reply src/components/SchemaSelect/schemaSelect.tsx:9]: schema select logic change

File: src/components/SchemaSelect/schemaSelect.types.ts [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+// Type guards to check if an object is a schemaTreeDir or schemaTreeItem
+function isSchemaTreeDir(item: schemaTreeDir | schemaTreeItem): item is schemaTreeDir {
+  return item && typeof item === 'object' && 'children' in item;
+}
+
+function isSchemaTreeItem(item: schemaTreeDir | schemaTreeItem): item is schemaTreeItem {
+  return item && typeof item === 'object' && 'id' in item;
+}
```
[the remark this answers is not in the document]
Comment (almog8k) [reply src/components/SchemaSelect/schemaSelect.types.ts:37]: changed

File: src/components/SchemaSelect/schemaSelect.tsx [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+        getOptionLabel={(option) => option.name ?? ''}
+        renderInput={(params) => <TextField {...params} label="Select Mid Level" />}
+        onChange={handleMidLevelChange}
+        value={midLevelSelection}
+        sx={{ width: 500 }}
+        disabled={!topLevelSelection}
+      />
+
+      <Autocomplete
+        id="grouped-demo"
```
[the remark this answers is not in the document]
Comment (almog8k) [reply src/components/SchemaSelect/schemaSelect.tsx:66]: changed

File: src/pages/config/viewConfigJson/jsonFormatter.ts [hunk 7]
```diff
@@ -1,3 +1,6 @@
-export const jsonFormatter = (data: unknown) => {
+export const jsonFormatter = (data: unknown | undefined) => {
+  if (!data) {
+    return '';
+  }
   return JSON.stringify(JSON.parse(JSON.stringify(data)), null, 2);
```
[the remark this answers is not in the document]
Comment (almog8k) [reply src/pages/config/viewConfigJson/jsonFormatter.ts:5]: simplified

File: src/pages/createConfig/createConfig.tsx [hunk 8]
```diff
@@ -1,7 +1,121 @@
...
+import { useMutation } from '@tanstack/react-query';
+import { GeneralInfoForm } from './step1GeneralInfo/step1GeneralInfo.schemas';
+
 export const CreateConfigsPage: React.FC = () => {
+  const [currentStep, setCurrentStep] = useState<StepEnum>(StepEnum.STEP1);
+  const [step1Data, setStep1Data] = useState<GeneralInfoForm | undefined>(undefined);
+  const [isStep1Valid, setIsStep1Valid] = useState<boolean>(false);
+  const [step2Data, setStep2Data] = useState<ConfigData>();
+  const [step2JsonStringData, setStep2JsonStringData] = useState<string | undefined>(undefined);
+  const [isStep2Valid, setIsStep2Valid] = useState<boolean>(false);
```
[the remark this answers is not in the document]
Comment (almog8k) [reply src/pages/createConfig/createConfig.tsx:20]: changed

File: src/pages/createConfig/createConfig.types.ts [hunk 9]
```diff
@@ -0,0 +1,12 @@
+import { config as Config } from '../../api/client/types.gen';
+
+export enum StepEnum {
+  STEP1 = 0,
+  STEP2 = 1,
+  STEP3 = 2,
+  FINISH = 3,
+}
```
[the remark this answers is not in the document]
Comment (almog8k) [reply src/pages/createConfig/createConfig.types.ts:8]: added comments

File: src/pages/createConfig/step2AddConfig/step2AddConfig.tsx (hunk 1 above)
Comment (almog8k) [reply src/pages/createConfig/step2AddConfig/step2AddConfig.tsx:28]: done

File: src/api/client/types.gen.ts [hunk 10]
```diff
@@ -8,7 +8,7 @@ export type configName = string;
 
 export type schemaId = string;
 
-export type version = number;
+export type version = number | 'latest';
```
Comment (CptSchnitz) [inline src/api/client/types.gen.ts:11]: dont change auto generated types.

**Review by CptSchnitz — CHANGES_REQUESTED** [review]

File: src/components/SchemaSelect/schemaSelect.tsx [hunk 11]
```diff
@@ -1,85 +1,50 @@
...
-        )}
-        disabled={!midLevelSelection}
-        isOptionEqualToValue={(option, value) => option.id === value?.id}
-        onChange={handleLowLevelChange}
-      />
-    </Box>
+    <Autocomplete
+      id="select-schema"
+      value={initialSelection}
+      options={options.sort((a, b) => -b.group.localeCompare(a.group))}
```
[the remark this answers is not in the document]
Comment (almog8k) [reply src/components/SchemaSelect/schemaSelect.tsx]: moved

File: src/api/client/types.gen.ts (hunk 10 above)
Comment (almog8k) [reply src/api/client/types.gen.ts]: revert

**Review by CptSchnitz — APPROVED** [review]

**Pull request merged**